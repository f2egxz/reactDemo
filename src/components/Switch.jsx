//-*- mode: react -*-
//-*- coding: utf-8 -*-

/**
 * Switch.jsx
 *
 * -TODO 小中大三种大小的开关按钮
 *
 *
 * @author fe.gxz
 */

import React          from 'react';
import { classnames } from 'util/view-extra'
import Styles         from 'style/Switch.css'


function BoxBoundClassNamesBySize (type,bool,disabled){
  return classnames(type, bool?Styles.switchBoxActive:'',disabled?Styles.disabledBox:'')
}

function BoxMapSizeToClassNames(size,bool,disabled){
  switch (size){
    case 'big':
      return BoxBoundClassNamesBySize(Styles.BSwitchBox,bool,disabled)
    case 'default':
      return BoxBoundClassNamesBySize(Styles.SwitchBox,bool,disabled)
    case 'small':
      return BoxBoundClassNamesBySize(Styles.SSwitchBox,bool,disabled)
    default:
      throw new TypeError(`Unknow Switch size ${size}`)
  }
}

function ItemBoundClassNamesBySize(item,itemActive,bool,disabled){
  return classnames(item,bool?itemActive:'',disabled?Styles.disabledItem:'')
}

function ItemMapSizeToClassNames(size,bool,disabled) {
  switch (size){
    case 'big':
      return ItemBoundClassNamesBySize(Styles.BSwitchItem,Styles.BSwitchItemActive,bool,disabled)
    case 'default':
      return ItemBoundClassNamesBySize(Styles.SwitchItem,Styles.SwitchItemActive,bool,disabled)
    case 'small':
      return ItemBoundClassNamesBySize(Styles.SSwitchItem,Styles.SSwitchItemActive,bool,disabled)
    default:
      throw new TypeError(`Unknow Switch size ${size}`)
  }
}

function boundClick (handleClick,disabled){
  return ()=>{
    if(!disabled){
      handleClick();
    }
  }
}

/**
 * Switch
 *
 * 开关
 * @param {boolean} bool
 * @param {string} size
 * @param {boolean} disabled
 * @param {func} handleClick
 * @return {html} switch
 */
function Switch({
  bool,
  size,
  disabled,
  handleClick
}) {
  return (
    <div className= {BoxMapSizeToClassNames(size,bool,disabled)} onClick= {boundClick(handleClick,disabled)}>
      <span className = {ItemMapSizeToClassNames(size,bool,disabled)}/>
    </div>
  )
}

Switch.propTypes = {
  bool       : React.PropTypes.bool.isRequired,
  disabled   : React.PropTypes.bool,
  handleClick: React.PropTypes.func.isRequired,
  size: React.PropTypes.oneOf([
    'small',
    'default',
    'big'
  ])
}

Switch.defaultProps = {
  size:'default',
  disabled:false
}


export default Switch
