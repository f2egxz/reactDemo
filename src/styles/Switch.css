/*-*- mode: css -*-*/
/*-*- coding: utf-8 -*-*/

/**
 * Switch.css
 *
 * 开关
 *
 * @author fe.gxz
 */

/**
 * Switch size
 */
@value itemSmall   1rem;
@value itemDefault 1.5rem;
@value itemBig     2rem;



/**
 * Switch theme color
 */

@value bgc           #c0c0c0;
@value bgcActive     #88C9F6;


/**
 * SwitchBoxTheme
 * @mixin
 * @param {size}  $itemSize
 * @param {color} $bgc
 */

@define-mixin switchBoxTheme
$BoxSize,
$bgc {

  position: relative;
  height: $BoxSize;
  width: calc($BoxSize*2);
  border-radius: calc($BoxSize/2);
  background-color: $bgc;
  border: 2px solid $bgc;

&:focus {
   box-shadow: 0 0 0 2px rgba(16, 142, 233, 0.2);
   outline: 0;
 }
}


/**
 * SwtichItemTheme
 * @mixin
 * @param {size} itemSize
 */
@define-mixin switchItemTheme
$itemSize {

  position: absolute;
  left:0;
  top:0;
  size:$itemSize;
  border-radius: calc($itemSize/2);
  background-color: #fff;
  transition: all 0.4s;

&:active {
   transform: scaleX(1.2);
 }
}

/**
 * SwtichItemActiveTheme
 * @mixin
 * @param {size} $itemSize
 */
@define-mixin switchItemActiveTheme
$itemSize {
  left: calc(100% - $itemSize);
}



.BSwitchBox {
  @mixin switchBoxTheme
  itemBig,
  bgc
}
.SwitchBox {
  @mixin switchBoxTheme
  itemDefault,
  bgc
}
.SSwitchBox {
  @mixin switchBoxTheme
  itemSmall,
  bgc
}

.switchBoxActive {
  background-color: bgcActive;
  border: 2px solid bgcActive;
}

.BSwitchItem {
  @mixin switchItemTheme
  itemBig
}
.BSwitchItemActive {
  @mixin switchItemActiveTheme
  itemBig
}


.SwitchItem {
  @mixin switchItemTheme
  itemDefault
}

.SwitchItemActive {
  @mixin switchItemActiveTheme
  itemDefault
}


.SSwitchItem {
  @mixin switchItemTheme
  itemSmall
}

.SSwitchItemActive {
  @mixin switchItemActiveTheme
  itemSmall
}

.disabledBox {
  cursor: not-allowed;
  background: #f4f4f4;
  border-color: #f4f4f4;
}
.disabledItem {
  background: #ccc;
  cursor: not-allowed;
}
