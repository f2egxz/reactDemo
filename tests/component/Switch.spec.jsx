import React from 'react'
import Switch from 'component/Switch.jsx'
import render from 'react-test-renderer'

test('click chanage the Switch',()=>{
  const SwitchClick= jest.fn();

  const component = render.create(
    <Switch bool={true} handleClick={SwitchClick}/>
  );

  let SwitchString = component.toJSON();
  expect(SwitchString).toMatchSnapshot();

  // component.props.onClick()
  // SwitchString = component.toJSON();
  //
  // expect(SwitchString).toMatchSnapshot();
})
