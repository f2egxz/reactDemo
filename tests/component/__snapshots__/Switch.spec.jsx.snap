exports[`test click chanage the Switch 1`] = `
<div
  className="SwitchBox switchBoxActive "
  onClick={[Function]}>
  <span
    className="SwitchItem SwitchItemActive " />
</div>
`;
